source/Ship.cpp

```diff
@@ -2645,12 +2739,158 @@ bool Ship::IsParalyzed() const
+	//double currentEnergy = GetCurrentEnergy();
+	if(totalConsumption < 0.)
+	{
+		double secondsToEmpty = currentEnergy / (-totalConsumption);
+		// how quickly can it charge under no energy load
+		double secondsToFullCharge = (idleEnergy > 0.) ? (maxEnergy - currentEnergy) / idleEnergy : minimumOperatingTime + 1.;
+		if((secondsToEmpty < minimumOperatingTime && secondsToFullCharge > minimumOperatingTime) || (secondsToEmpty < minimumOperatingTime && idleEnergy <= 0.))
+			return true;
+	}
+	return false;
+}
+
+
+
+// Check if escort fleet is full of fuel.  In this case, the fighters can help
+// replenish the Carrier fuel tanker.
+bool Ship::IsEscortsFullOfFuel() const
+{
+	if(!IsYours())
+		return false;
+	return isEscortsFullOfFuel;
+}
+
+
+
+// Out of energy ships will be disabled.  Out of energy fighters will be
+// disabled if far away from the parent.  This is so the parent can board the
+// fighter however the fighter needs to not be disabled in order to board the
+// carrier.
+bool Ship::IsOutOfEnergy() const
+{
+	bool closeToParent = false;
+	bool outOfEnergy = GetCurrentEnergy() <= 0. && GetMovingEnergyPerFrame() > 0.  && GetIdleEnergyPerFrame() <= 0.;
+	if(canBeCarried && outOfEnergy)
+	{
+		if(GetParent())
+		{
+			Point dp = GetParent().get()->Position() - position;
+			Point dv = GetParent().get()->Velocity() - velocity;
+			double distanceFromParent = dp.Length();
+			double speedRelativeToParent = dv.Length();
+			closeToParent = distanceFromParent < 50. && speedRelativeToParent < 1.;
+			// Set closeToParent to false if this ship is disabled due to hull.  It requires additional assistance.
+			closeToParent &= hull >= minimumHull;
+			closeToParent &= static_cast<bool>(GetSystem());
+			// CanCarry is expensive performance-wise so should only be called if necessary.
+			if(closeToParent)
+				closeToParent &= GetParent()->BaysFree(attributes.Category());
+			// Clear the parent boarding target when fighter boards.
+			if(closeToParent && GetParent().get()->GetTargetShip().get() == this)
+				GetParent().get()->SetTargetShip(nullptr);
+		}
+		else
+			closeToParent = false;
+	}
+	// Return will always be false if not a fighter.
+	return !landingPlanet && !hasBoarded && !closeToParent && outOfEnergy;
+}
+
+
+
+// Check if ship fuel is low or check destination ship fuel (compareTo) can refuel.
+bool Ship::IsFuelLow() const
+{
+	return IsFuelLow(attributes.Get("fuel capacity"));
+}
+
+
+
+// Check if ship fuel is low or check destination ship fuel (compareTo) can
+// refuel.  If the ship has no fuel then it can't be low.
+bool Ship::IsFuelLow(double compareTo) const
+{
+	bool lowFuel = attributes.Get("fuel capacity");
+	// Fighters can have less than 100 fuel and can also transfer 100% of fuel
+	// to other escorts.
+	if(CanBeCarried())
+		lowFuel &= (fuel < 100. || Fuel() < .5) && Fuel() < 1.;
+	else if(IsYours() || GetPersonality().IsEscort())
+		lowFuel &= fuel < attributes.Get("fuel capacity");
```

Comment:
~~How dare you query fuel capacity twice~~
Seriously though, aint any need for that.

Btw considering how many times we're querying it you could maybe consider putting it in a quick variable that gets updated only when you changed outfits with an outfit that has fuel capacity/finish loading the ship?